MemoryManagement/MemoryManagement/MSKContactDetailViewController.m

```diff
@@ -0,0 +1,64 @@
+//
+//  MSKContactDetailViewController.m
+//  MemoryManagement
+//
+//  Created by Lambda_School_Loaner_268 on 5/20/20.
+//  Copyright © 2020 Lambda. All rights reserved.
+//
+
+#import "MSKContactDetailViewController.h"
+#import "MSKContact.h"
+#import "ContactTableViewController.h"
+#import "MainNav.h"
+
+
+@interface MSKContactDetailViewController ()
+@end
+
+@implementation MSKContactDetailViewController
+-(instancetype)init {
+    self = [super init];
+    return self;
+}
+- (void)dealloc {
+    [_nameTF release];
+    [_phoneTf release];
+    [_emailTF release];
+    [super dealloc];
+}
+- (IBAction)save:(UIBarButtonItem *)sender {
+    if (_nameTF == nil) {
+        self.nameTF.text = @"Unknown Name";
+    }
+
+    [contacts addObject:[[MSKContact alloc] initWithName:self.nameTF.text
+                                             contactEmail:self.emailTF.text
+                                             contactPhone:self.phoneTf.text]];
+    [self dealloc];
+    ContactTableViewController *bad = [[ContactTableViewController alloc] init];
+    [self presentViewController:bad  animated:true completion:nil];
+    }
```

Comment:
A couple of things here:

- Remember we should use `self.nameTF` (etc) rather than `_nameTF` everywhere except in init and dealloc, or else we can end up with some unexpected behavior. `self.property` will call the getter/setter methods, which may perform some other logic for us, while `_property` directly accesses the reference to the object, bypassing any setter/getter. Usually we don't want this, as the getter and setter may be doing something important.
- `nameTF` shouldn't ever be nil when we hit the save button. The `text` on the textField might be `nil` or equal to `@""`; that's what we want to check.
- If we `dealloc` self and then try to use it to present a view controller, we're going to have a bad time. In addition, presenting a new view controller is going to end us up in a recursive loop where we keep pushing and presenting new view controllers. Instead, we can use the navigationController of this viewController to pop this view controller off of the navigation stack. We shouldn't need to manually dealloc at all.

```objc
- (IBAction)save:(UIBarButtonItem *)sender {
    if (self.nameTF.text == nil || [self.nameTF.text isEqual:@""]) {
        self.nameTF.text = @"Unknown Name";
    }

    [contacts addObject:[[MSKContact alloc] initWithName:self.nameTF.text
                                            contactEmail:self.emailTF.text
                                            contactPhone:self.phoneTf.text]];
    [self.navigationController popViewControllerAnimated:YES];
}
```

Reply:
Thanks for the feedback. I am sorry that I don't respond to these, but I always read them. You're a much better dev than me, wonder whether that's been having to teach it, effort, or innate skill?